// @elsdoerfer/react-arrow/index.es.js
import _extends from 'babel-runtime/helpers/extends';
import _objectWithoutProperties from 'babel-runtime/helpers/objectWithoutProperties';
import _Object$getPrototypeOf from 'babel-runtime/core-js/object/get-prototype-of';
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _createClass from 'babel-runtime/helpers/createClass';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';
import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

// http://thenewcode.com/1068/Making-Arrows-in-SVG

var PolyLineStyle = {
  fill: 'none'
  // Not yet sure if this is a good property to set or not
  //vectorEffect: 'non-scaling-stroke'
};

var PathStyle = {
  fill: 'none'
};

function toRad(d) {
  return d * Math.PI / 180;
}

var Arrow = function (_PureComponent) {
  _inherits(Arrow, _PureComponent);

  function Arrow() {
    _classCallCheck(this, Arrow);

    var _this = _possibleConstructorReturn(this, (Arrow.__proto__ || _Object$getPrototypeOf(Arrow)).call(this));

    if (!_this.constructor.__counter) {
      _this.constructor.__counter = 1;
    }
    _this.uniqid = _this.constructor.__counter++;
    return _this;
  }

  _createClass(Arrow, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          length = _props.length,
          angle = _props.angle,
          arrowHeadFilled = _props.arrowHeadFilled,
          lineWidth = _props.lineWidth,
          lineDashed = _props.lineDashed,
          otherProps = _objectWithoutProperties(_props, ['length', 'angle', 'arrowHeadFilled', 'lineWidth', 'lineDashed']);

      // By default, our sin math would let the angle rotate to
      // the left. Reverse the direction.


      angle = -angle + 180;
      angle = angle % 360;

      // Let's do some trig to calculate the viewbox.
      //
      // Assuming the arrow points top right, c is the arrow
      // length, a is the viewbox width, b is the viewbox height.

      // We know C is 90deg
      var B = angle;
      var C = 90;
      var A = 180 - C - B;
      var c = length;

      // Get all the sides
      var a = c / Math.sin(toRad(C)) * Math.sin(toRad(A));
      var b = c / Math.sin(toRad(C)) * Math.sin(toRad(B));

      // a, b can be negative if the angle. We will rewrite
      // those values to go inside the svg viewbox.
      var width = Math.abs(b);
      var height = Math.abs(a);

      var padding = 10;

      // Now we have a viewBox
      var viewBox = '0 0 ' + (width + padding * 2) + ' ' + (height + padding * 2);

      function point(x, y) {
        if (b < 0) {
          x = width + padding * 2 - x;
        }
        if (a < 0) {
          y = height + padding * 2 - y;
        }
        return x + ',' + y;
      }

      var path = ['M' + point(padding, padding), 'L' + point(width + padding, height + padding)];

      // TODO: Cooler line:
      // M16.7,178 c87.6-46.9,162.9-185.4,227-136.4C307.2,90.1,195,158.5,111,108.9C71,85.2,92.2,30.7,126,7

      var markerId = 'Arrow-pointer-' + this.uniqid;

      // Based on the props, determine the styles.
      var arrowHeadStyle = _extends({}, PolyLineStyle, {
        strokeWidth: 0,
        stroke: this.props.color
      });
      if (this.props.arrowHeadFilled) {
        arrowHeadStyle.fill = this.props.color;
      } else {
        arrowHeadStyle.strokeWidth = 1;
      }

      var lineStyle = _extends({}, PathStyle, {
        stroke: this.props.color,
        strokeWidth: this.props.lineWidth
      });

      if (lineDashed) {
        if (typeof lineDashed !== 'string') {
          lineDashed = '11, 5';
        }

        lineStyle.strokeDasharray = lineDashed;
      }

      return React.createElement(
        'svg',
        _extends({ xmlns: 'http://www.w3.org/2000/svg', viewBox: viewBox }, otherProps),
        React.createElement(
          'defs',
          null,
          React.createElement(
            'marker',
            {
              id: markerId,
              markerWidth: '9',
              markerHeight: '9',
              refX: '8',
              refY: '5',
              orient: 'auto',
              markerUnits: 'strokeWidth'
            },
            React.createElement('polyline', { points: '1 1, 9 5, 1 9', style: arrowHeadStyle })
          )
        ),
        React.createElement('path', {
          style: lineStyle,
          d: path.join(' '),
          markerEnd: 'url(#' + markerId + ')'
        })
      );
    }
  }]);

  return Arrow;
}(PureComponent);

Arrow.propTypes = {
  length: PropTypes.number,
  angle: PropTypes.number,
  color: PropTypes.string,
  arrowHeadFilled: PropTypes.bool,
  lineWidth: PropTypes.number,
  lineDashed: PropTypes.oneOfType([PropTypes.bool, PropTypes.string])
};
Arrow.defaultProps = {
  angle: 0,
  length: 50,
  color: '#231F20',
  arrowHeadFilled: true,
  lineWidth: 1,
  lineDashed: false
};

export default Arrow;
